File: examples/xhr/src/xhr.js
export function simpleGet(port) {
    console.log('XMLHttpRequest test 1: simple GET');
    
    return new Promise((resolve) => {
        const xhr = new XMLHttpRequest();
        
        xhr.onreadystatechange = function() {
            if (xhr.readyState === xhr.DONE) {
                console.log(`Status: ${xhr.status}`);
                console.log(`Response length: ${xhr.response ? xhr.response.length : 0}`);
                resolve();
            }
        };
        
        xhr.open('GET', `http://localhost:${port}/todos`);
        xhr.send();
    });
}

export function simplePost(port) {
    console.log('XMLHttpRequest test 2: simple POST');
    
    return new Promise((resolve) => {
        const xhr = new XMLHttpRequest();
        
        xhr.onload = function() {
            console.log(`POST Status: ${xhr.status}`);
            console.log(`POST Response length: ${xhr.response ? xhr.response.length : 0}`);
            resolve();
        };
        
        xhr.onerror = function() {
            console.log('POST Error occurred');
            resolve();
        };
        
        xhr.open('POST', `http://localhost:${port}/todos`);
        xhr.setRequestHeader('Content-Type', 'application/json');
        xhr.send(JSON.stringify({
            title: 'Test Todo',
            body: 'Testing XMLHttpRequest',
            userId: 1
        }));
    });
}

export function setRequestHeaders(port) {
    console.log('XMLHttpRequest test 3: set request headers');
    
    return new Promise((resolve) => {
        const xhr = new XMLHttpRequest();
        
        xhr.onload = function() {
            console.log(`Headers test status: ${xhr.status}`);
            resolve();
        };
        
        xhr.open('GET', `http://localhost:${port}/todos`);
        xhr.setRequestHeader('X-Custom-Header', 'test-value');
        xhr.setRequestHeader('Accept', 'application/json');
        xhr.send();
    });
}

export function getResponseHeaders(port) {
    console.log('XMLHttpRequest test 4: get response headers');
    
    return new Promise((resolve) => {
        const xhr = new XMLHttpRequest();
        
        xhr.onload = function() {
            const contentType = xhr.getResponseHeader('content-type');
            console.log(`Content-Type header: ${contentType}`);
            
            const allHeaders = xhr.getAllResponseHeaders();
            console.log(`All headers length: ${allHeaders.length}`);
            
            resolve();
        };
        
        xhr.open('GET', `http://localhost:${port}/todos`);
        xhr.send();
    });
}

export function responseTypes(port) {
    console.log('XMLHttpRequest test 5: response types');
    
    return new Promise(async (resolve) => {
        // Test text response
        const xhr1 = new XMLHttpRequest();
        xhr1.responseType = 'text';
        
        xhr1.onload = function() {
            console.log(`Text response type: ${typeof xhr1.response}`);
            
            // Test JSON response
            const xhr2 = new XMLHttpRequest();
            xhr2.responseType = 'json';
            
            xhr2.onload = function() {
                console.log(`JSON response type: ${typeof xhr2.response}`);
                console.log(`JSON response is object: ${typeof xhr2.response === 'object'}`);
                resolve();
            };
            
            xhr2.open('GET', `http://localhost:${port}/todos`);
            xhr2.send();
        };
        
        xhr1.open('GET', `http://localhost:${port}/todos`);
        xhr1.send();
    });
}

export function readystateEvents(port) {
    console.log('XMLHttpRequest test 6: readystate events');
    
    return new Promise((resolve) => {
        const xhr = new XMLHttpRequest();
        const states = [];
        
        xhr.onreadystatechange = function() {
            states.push(xhr.readyState);
            console.log(`ReadyState changed to: ${xhr.readyState}`);
        };
        
        xhr.onload = function() {
            console.log(`States encountered: ${states.join(', ')}`);
            console.log(`Final status: ${xhr.status}`);
            resolve();
        };
        
        xhr.open('GET', `http://localhost:${port}/todos`);
        xhr.send();
    });
}

export function errorHandling(port) {
    console.log('XMLHttpRequest test 7: error handling');
    
    return new Promise((resolve) => {
        const xhr = new XMLHttpRequest();
        
        xhr.onerror = function() {
            console.log('Error handler called');
            console.log(`Status on error: ${xhr.status}`);
            resolve();
        };
        
        xhr.onload = function() {
            console.log('Request succeeded');
            resolve();
        };
        
        // Try to fetch from invalid URL to trigger error
        xhr.open('GET', 'http://invalid-domain-that-does-not-exist-12345.com/test');
        xhr.send();
    });
}

export function abortRequest(port) {
    console.log('XMLHttpRequest test 8: abort request');
    
    return new Promise((resolve) => {
        const xhr = new XMLHttpRequest();
        let abortCalled = false;
        
        xhr.onabort = function() {
            console.log('Abort handler called');
            abortCalled = true;
            resolve();
        };
        
        xhr.onload = function() {
            console.log('Request completed (should not happen)');
            resolve();
        };
        
        xhr.open('GET', `http://localhost:${port}/todos-stream`);
        xhr.send();
        
        // Abort immediately after send
        setTimeout(() => {
            xhr.abort();
        }, 10);
    });
}

export function timeoutHandling(port) {
    console.log('XMLHttpRequest test 9: timeout handling');
    
    return new Promise((resolve) => {
        const xhr = new XMLHttpRequest();
        let timeoutCalled = false;
        
        xhr.timeout = 100; // 100ms timeout
        
        xhr.ontimeout = function() {
            console.log('Timeout handler called');
            timeoutCalled = true;
            resolve();
        };
        
        xhr.onload = function() {
            console.log('Request completed');
            if (!timeoutCalled) {
                resolve();
            }
        };
        
        // Use a slow endpoint to trigger timeout
        xhr.open('GET', `http://localhost:${port}/slow`);
        xhr.send();
    });
}

export function requestWithBasicAuth(port) {
    console.log('XMLHttpRequest test 10: basic auth');
    
    return new Promise((resolve) => {
        const xhr = new XMLHttpRequest();
        
        xhr.onload = function() {
            console.log(`Auth request status: ${xhr.status}`);
            resolve();
        };
        
        xhr.onerror = function() {
            console.log('Auth request error');
            resolve();
        };
        
        // Basic auth is set via open method
        xhr.open('GET', `http://localhost:${port}/todos`, true, 'testuser', 'testpass');
        xhr.send();
    });
}

export function postWithFormData(port) {
    console.log('XMLHttpRequest test 11: POST with FormData');
    
    return new Promise((resolve) => {
        const xhr = new XMLHttpRequest();
        const formData = new FormData();
        
        formData.append('name', 'John Doe');
        formData.append('email', 'john@example.com');
        
        xhr.onload = function() {
            console.log(`FormData POST status: ${xhr.status}`);
            resolve();
        };
        
        xhr.open('POST', `http://localhost:${port}/form-echo`);
        xhr.send(formData);
    });
}

export function postWithJsonBody(port) {
    console.log('XMLHttpRequest test 12: POST with JSON body');
    
    return new Promise((resolve) => {
        const xhr = new XMLHttpRequest();
        
        const jsonData = {
            name: 'Test',
            value: 42
        };
        
        xhr.onload = function() {
            console.log(`JSON POST status: ${xhr.status}`);
            console.log(`JSON response: ${xhr.response}`);
            resolve();
        };
        
        xhr.open('POST', `http://localhost:${port}/json-echo`);
        xhr.setRequestHeader('Content-Type', 'application/json');
        xhr.send(JSON.stringify(jsonData));
    });
}

export function statusIsNumber(port) {
    console.log('XMLHttpRequest test 13: status is number');
    
    return new Promise((resolve) => {
        const xhr = new XMLHttpRequest();
        
        xhr.onload = function() {
            const status = xhr.status;
            console.log(`Status value: ${status}`);
            console.log(`Status typeof: ${typeof status}`);
            console.log(`Status is number: ${typeof status === 'number'}`);
            console.log(`Status === 200: ${status === 200}`);
            console.log(`Status >= 200 && status < 300: ${status >= 200 && status < 300}`);
            resolve();
        };
        
        xhr.onerror = function() {
            console.log('Error occurred');
            resolve();
        };
        
        xhr.open('GET', `http://localhost:${port}/todos`);
        xhr.send();
    });
}

File: tests/runtime/xhr.rs
use crate::common::test_server::start_test_server;
use crate::common::{CompiledTest, invoke_and_capture_output};
use camino::Utf8Path;
use test_r::{test, test_dep};
use wasmtime::component::Val;

#[test_dep(tagged_as = "xhr")]
fn compiled_xhr() -> CompiledTest {
    let path = Utf8Path::new("examples/xhr");
    CompiledTest::new(path, true).expect("Failed to compile xhr")
}

#[test]
async fn xhr_simple_get(#[tagged_as("xhr")] compiled: &CompiledTest) -> anyhow::Result<()> {
    let (port, _) = start_test_server().await;

    let (r, output) =
        invoke_and_capture_output(compiled.wasm_path(), None, "simple-get", &[Val::U16(port)])
            .await;
    let _ = r?;

    println!("{output}");

    assert!(output.contains("XMLHttpRequest test 1: simple GET"));
    assert!(output.contains("Status: 200"));
    assert!(output.contains("Response length:"));

    Ok(())
}

#[test]
async fn xhr_simple_post(#[tagged_as("xhr")] compiled: &CompiledTest) -> anyhow::Result<()> {
    let (port, _) = start_test_server().await;

    let (r, output) =
        invoke_and_capture_output(compiled.wasm_path(), None, "simple-post", &[Val::U16(port)])
            .await;
    let _ = r?;

    println!("{output}");

    assert!(output.contains("XMLHttpRequest test 2: simple POST"));
    assert!(output.contains("POST Status:") || output.contains("POST Error occurred"));

    Ok(())
}

#[test]
async fn xhr_set_request_headers(
    #[tagged_as("xhr")] compiled: &CompiledTest,
) -> anyhow::Result<()> {
    let (port, _) = start_test_server().await;

    let (r, output) = invoke_and_capture_output(
        compiled.wasm_path(),
        None,
        "set-request-headers",
        &[Val::U16(port)],
    )
    .await;
    let _ = r?;

    println!("{output}");

    assert!(output.contains("XMLHttpRequest test 3: set request headers"));
    assert!(output.contains("Headers test status:") || output.contains("ReadyState changed"));

    Ok(())
}

#[test]
async fn xhr_get_response_headers(
    #[tagged_as("xhr")] compiled: &CompiledTest,
) -> anyhow::Result<()> {
    let (port, _) = start_test_server().await;

    let (r, output) = invoke_and_capture_output(
        compiled.wasm_path(),
        None,
        "get-response-headers",
        &[Val::U16(port)],
    )
    .await;
    let _ = r?;

    println!("{output}");

    assert!(output.contains("XMLHttpRequest test 4: get response headers"));
    assert!(output.contains("Content-Type header:") || output.contains("All headers length:"));

    Ok(())
}

#[test]
async fn xhr_response_types(#[tagged_as("xhr")] compiled: &CompiledTest) -> anyhow::Result<()> {
    let (port, _) = start_test_server().await;

    let (r, output) = invoke_and_capture_output(
        compiled.wasm_path(),
        None,
        "response-types",
        &[Val::U16(port)],
    )
    .await;
    let _ = r?;

    println!("{output}");

    assert!(output.contains("XMLHttpRequest test 5: response types"));
    assert!(output.contains("Text response type:") || output.contains("JSON response type:"));

    Ok(())
}

#[test]
async fn xhr_readystate_events(#[tagged_as("xhr")] compiled: &CompiledTest) -> anyhow::Result<()> {
    let (port, _) = start_test_server().await;

    let (r, output) = invoke_and_capture_output(
        compiled.wasm_path(),
        None,
        "readystate-events",
        &[Val::U16(port)],
    )
    .await;
    let _ = r?;

    println!("{output}");

    assert!(output.contains("XMLHttpRequest test 6: readystate events"));
    assert!(output.contains("ReadyState changed to:") || output.contains("States encountered:"));

    Ok(())
}

#[test]
async fn xhr_error_handling(#[tagged_as("xhr")] compiled: &CompiledTest) -> anyhow::Result<()> {
    let (port, _) = start_test_server().await;

    let (r, output) = invoke_and_capture_output(
        compiled.wasm_path(),
        None,
        "error-handling",
        &[Val::U16(port)],
    )
    .await;
    let _ = r?;

    println!("{output}");

    assert!(output.contains("XMLHttpRequest test 7: error handling"));
    // Either error handler is called or request succeeds
    assert!(
        output.contains("Error handler called")
            || output.contains("Request succeeded")
            || output.contains("Status on error:")
    );

    Ok(())
}

#[test]
async fn xhr_abort_request(#[tagged_as("xhr")] compiled: &CompiledTest) -> anyhow::Result<()> {
    let (port, _) = start_test_server().await;

    let (r, output) = invoke_and_capture_output(
        compiled.wasm_path(),
        None,
        "abort-request",
        &[Val::U16(port)],
    )
    .await;
    let _ = r?;

    println!("{output}");

    assert!(output.contains("XMLHttpRequest test 8: abort request"));
    assert!(output.contains("Abort handler called") || output.contains("Request completed"));

    Ok(())
}

#[test]
async fn xhr_timeout_handling(#[tagged_as("xhr")] compiled: &CompiledTest) -> anyhow::Result<()> {
    let (port, _) = start_test_server().await;

    let (r, output) = invoke_and_capture_output(
        compiled.wasm_path(),
        None,
        "timeout-handling",
        &[Val::U16(port)],
    )
    .await;
    let _ = r?;

    println!("{output}");

    assert!(output.contains("XMLHttpRequest test 9: timeout handling"));
    // Either timeout or normal completion
    assert!(output.contains("Timeout handler called") || output.contains("Request completed"));

    Ok(())
}

#[test]
async fn xhr_post_with_form_data(
    #[tagged_as("xhr")] compiled: &CompiledTest,
) -> anyhow::Result<()> {
    let (port, _) = start_test_server().await;

    let (r, output) = invoke_and_capture_output(
        compiled.wasm_path(),
        None,
        "post-with-form-data",
        &[Val::U16(port)],
    )
    .await;
    let _ = r?;

    println!("{output}");

    assert!(output.contains("XMLHttpRequest test 11: POST with FormData"));
    assert!(output.contains("FormData POST status:") || output.contains("ReadyState changed"));

    Ok(())
}

#[test]
async fn xhr_post_with_json_body(
    #[tagged_as("xhr")] compiled: &CompiledTest,
) -> anyhow::Result<()> {
    let (port, _) = start_test_server().await;

    let (r, output) = invoke_and_capture_output(
        compiled.wasm_path(),
        None,
        "post-with-json-body",
        &[Val::U16(port)],
    )
    .await;
    let _ = r?;

    println!("{output}");

    assert!(output.contains("XMLHttpRequest test 12: POST with JSON body"));
    assert!(output.contains("JSON POST status:") || output.contains("ReadyState changed"));

    Ok(())
}

#[test]
async fn xhr_basic_auth(#[tagged_as("xhr")] compiled: &CompiledTest) -> anyhow::Result<()> {
    let (port, _) = start_test_server().await;

    let (r, output) = invoke_and_capture_output(
        compiled.wasm_path(),
        None,
        "request-with-basic-auth",
        &[Val::U16(port)],
    )
    .await;
    let _ = r?;

    println!("{output}");

    assert!(output.contains("XMLHttpRequest test 10: basic auth"));
    assert!(output.contains("Auth request status:") || output.contains("Auth request error"));

    Ok(())
}

#[test]
async fn xhr_status_is_number(#[tagged_as("xhr")] compiled: &CompiledTest) -> anyhow::Result<()> {
    let (port, _) = start_test_server().await;

    let (r, output) = invoke_and_capture_output(
        compiled.wasm_path(),
        None,
        "status-is-number",
        &[Val::U16(port)],
    )
    .await;
    let _ = r?;

    println!("{output}");

    assert!(output.contains("XMLHttpRequest test 13: status is number"));
    assert!(output.contains("Status typeof: number"));
    assert!(output.contains("Status is number: true"));
    assert!(output.contains("Status === 200: true"));
    assert!(output.contains("Status >= 200 && status < 300: true"));

    Ok(())
}
